[planning] Deprecate MaxCliqueSolverBase::Clone

jwnimmer-tri opened the pull request:
Amends #21892.  I don't understand why we offered cloning in the first place.  All of the base class methods are const.

Towards #21968.  Specifically, `PYBIND11_OVERRIDE_...(std::unique_ptr<...>, ...)` is no longer an allowed concept (Python cannot return uniquely-owned objects).  There are complicated ways around this, but in this case dropping the feature seems easiest.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/RobotLocomotion/drake/22425)
<!-- Reviewable:end -->

jwnimmer-tri commented:
+@AlexandreAmice for feature review, please?

AlexandreAmice reviewed (commented):
In #21892 I explicitly chose to mark the method `MinCliqueCoverSolverBase::SolveMinCliqueCover` non-const since I know that some (most) algorithms (though none that we have implemented) would be randomized and therefore may need to mutate a member (e.g. a `RandomGenerator`). 

I see that in `MaxCliqueSolver::SolveMaxClique` I marked it as `const`even though the same considerations apply. I would honestly prefer to remove the `const` marker on that method.

Does this change your thoughts on whether the `Clone` method should be removed?

> Reviewed 10 of 10 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22425#-:-OG6s1yxBHZLmHjgLYcJ:b2g7588)* status: LGTM missing from assignee AlexandreAmice, needs platform reviewer assigned, needs at least two assigned reviewers



<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
It's a good question, I will give it some thought.  No particular hurry here, we still have a few weeks or more before this becomes critical path.

jwnimmer-tri commented:
> In #21892 I explicitly chose to mark the method `MinCliqueCoverSolverBase::SolveMinCliqueCover` non-const since I know that some (most) algorithms (though none that we have implemented) would be randomized and therefore may need to mutate a member (e.g. a `RandomGenerator`). 
>
> I see that in `MaxCliqueSolver::SolveMaxClique` I marked it as `const`even though the same considerations apply. I would honestly prefer to remove the `const` marker on that method.

Okay, that would be fine.  I won't change it here, but I've reworked the proposal here to allow you to change that down the road.

(It would probably better for the RNG to be a mutable in,out function argument than member state, though, like we do with MonteCarlo.)

> Does this change your thoughts on whether the `Clone` method should be removed?

Nope.  I just changed it from `shared_ptr<const Foo>` to `shared_ptr<Foo>`.  Essentially what's being passed is a function handle.  It probably could have just been a `std::function`, but a `shared_ptr<VirtualInterface>` is basically the same thing.  We don't want to place any demands on the _lifetime_ or _cleanup_ the function handle, we just want to be able to call it.

Allowing users to set up thread-safe pools of objects and function pointers is outside of the scope of these classes.  Cloning doesn't realistically help with that, anyway.

AlexandreAmice reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22425#-:-OGzpYqV5sBK6RTdREhl:bnfp4nl)* status: needs platform reviewer assigned, needs at least two assigned reviewers



<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
+@ggould-tri for platform review per schedule, please.

ggould-tri reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 7 of 10 files at r1, 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22425#-:-OH8Nue3EdG7oZZAVAcq:bnfp4nl)* status: <span>:shipit:</span> complete! all discussions resolved, LGTM from assignees ggould-tri(platform),AlexandreAmice



<!-- Sent from Reviewable.io -->
